shardingsphere-kernel/shardingsphere-transaction/shardingsphere-transaction-core/src/main/java/org/apache/shardingsphere/transaction/context/TransactionContextsBuilder.java

```diff
@@ -61,4 +75,102 @@ private TransactionRule getTransactionRule() {
         Optional<TransactionRule> transactionRule = globalRules.stream().filter(each -> each instanceof TransactionRule).map(each -> (TransactionRule) each).findFirst();
         return transactionRule.orElseGet(() -> new TransactionRule(new DefaultTransactionRuleConfigurationBuilder().build()));
     }
+    
+    private void generateNarayanaConfig(final TransactionRule transactionRule) {
+        Map<Object, Object> result;
+        if (transactionRule.getDefaultType() == TransactionType.XA && transactionRule.getProviderType().equalsIgnoreCase("Narayana")) {
+            result = generateDefaultNarayanaConfig();
+            if (null != transactionRule.getProps()) {
+                swapJdbcStore(transactionRule, result);
+            }
+        } else {
+            return;
+        }
+        String value = narayanaConfigMapToXml((LinkedHashMap<Object, Object>) result);
+        String path = ClassLoader.getSystemResource("").getPath();
+        System.out.println(path);
+        try (BufferedWriter bufferedWriter = Files.newBufferedWriter(Paths.get(path, "jbossts-properties.xml"))) {
+            bufferedWriter.write(value);
+            bufferedWriter.flush();
+        } catch (final IOException ex) {
+            log.error("generate narayana config file failed.");
+        }
+    }
```

Comment:
This is for common transaction process, could you try to consider about abstract a SPI and move NarayanaConfig into the SPI implementation?

Reply:
Config of transaction is `org.apache.shardingsphere.transaction.rule.TransactionRule`, all types of transactions share the `props` field.
```
public final class TransactionRule implements GlobalRule {
    
    private final TransactionType defaultType;
    
    private final String providerType;
    
    private final Properties props; 
    
    public TransactionRule(final TransactionRuleConfiguration ruleConfig) {
        defaultType = TransactionType.valueOf(ruleConfig.getDefaultType().toUpperCase());
        providerType = ruleConfig.getProviderType();
        props = ruleConfig.getProps();
    }
    
    @Override
    public String getType() {
        return TransactionRule.class.getSimpleName();
    }
}
```
`TransactionContextsBuilder` is the general entry for transaction engine initialization. In `TransactionContextsBuilder`, the transaction engine is initialized according to the configuration.

Reply:
What I mean is `narayana` should not occur in TransactionContextsBuilder class, which is common one, should not handle concrete transaction type impl

Reply:
ok